src/main.rs

```diff
@@ -15,36 +15,70 @@ use std::io::BufWriter;
 use std::io::Write;
 use std::path::Path;
 use std::time::Instant;
+use zstd::stream::read::Decoder as ZstdDecoder;
 
 extern crate blas;
-extern crate intel_mkl_src;
 extern crate half;
+extern crate intel_mkl_src;
 
 #[macro_use]
 extern crate nom;
 extern crate core;
 
-use fw::{cmdline, feature_buffer, logging_layer, regressor};
 use fw::cache::RecordCache;
 use fw::feature_buffer::FeatureBufferTranslator;
 use fw::hogwild::HogwildTrainer;
 use fw::model_instance::{ModelInstance, Optimizer};
 use fw::multithread_helpers::BoxedRegressorTrait;
 use fw::parser::VowpalParser;
-use fw::persistence::{new_regressor_from_filename, save_regressor_to_filename, save_sharable_regressor_to_filename};
+use fw::persistence::{
+    new_regressor_from_filename, save_regressor_to_filename, save_sharable_regressor_to_filename,
+};
 use fw::regressor::{get_regressor_with_weights, Regressor};
 use fw::serving::Serving;
 use fw::vwmap::VwNamespaceMap;
+use fw::{cmdline, feature_buffer, logging_layer, regressor};
 
 fn main() {
     logging_layer::initialize_logging_layer();
 
-    if let Err(e) = main2() {
+    if let Err(e) = main_fw_loop() {
         log::error!("Global error: {:?}", e);
         std::process::exit(1)
     }
 }
 
+fn create_buffered_input(input_filename: &str) -> Box<dyn BufRead> {
+    // Handler for different (or no) compression types
+    
+    let input = File::open(input_filename).expect("Could not open the input file.");
+
+    let input_format = Path::new(&input_filename)
+        .extension()
+        .and_then(|ext| ext.to_str())
+        .expect("Failed to get the file extension.");
+
+    match input_format {
+        "gz" => {
+            let gz_decoder = MultiGzDecoder::new(input);
+            let reader = io::BufReader::new(gz_decoder);
+            Box::new(reader)
+        }
+        "zst" => {
```

Comment:
We should add tests when adding a new input/output format

Reply:
Added, good point

Reply:
Github knows you didn't add tests, now that you did the build passes :)

Reply:
Unfortunately no, build issue was due most likely due to stack/runner conflict, put that on todos